### PR TITLE
Fix bug for "Recording inserts too many notes" #1224

### PR DESCRIPTION
**Pull request by oddtime**
Trying to fix that bug, waiting discussion on how the GUI should operate in this case (replace the note? add a note?)

The first commit was a simple fix: "always add another note".
The second commit "replace the current note if found." But it misses undo/redo action.

See the "TODO" in comments
Hope that helps

### CODE REVIEW

**Comment by oddtime** [comment]
@cme in order to overwrite a note while recording (I thought by changing some of its properties, the velocity for sure),
should I use an existing undo action class or write another class? Thanks

**Comment by cme** [comment]
The undo actions badly need refactoring. For the pattern editor I'm going to unify most of them to actions that basically replace whole notes at a time (with 'add' and 'delete' being subcases of that with null 'before' and 'after' notes respectively) with some wrinkles around the edges for preserving selections.

Using the existing undo actions (which will also make it easier to refactor later) it's probably best to use a pair of `addOrDeleteNoteAction`s to remove the overwritten note then insert the new one. But wrap them in a macro so they're undone/redone as a single user action.

Thanks!

File: src/core/Hydrogen.cpp [hunk 1]
```diff
@@ -2721,6 +2721,7 @@ void Hydrogen::addRealtimeNote(	int		instrument,
 			noteAction.b_noteExist = ( pNoteold ) ? true : false;
 
 			EventQueue::get_instance()->m_addMidiNoteVector.push_back(noteAction);
+			printf("pushed note\n");
```
[the remark this answers is not in the document]
Comment (oddtime) [reply src/core/Hydrogen.cpp:2724]: Thanks, I was not aware in fact XD

[the remark this answers is not in the document]
Comment (oddtime) [reply src/core/Hydrogen.cpp:2724]: Of course that was temporary

[the remark this answers is not in the document]
Comment (cme) [reply src/core/Hydrogen.cpp:2724]: > Thanks, I was not aware in fact XD

Then I'm glad I mentioned it :D You can do `qDebug() << pNoteold` right here and you'll see all the note properties, it's kinda awesome.

**Comment by oddtime** [comment]
> Using the existing undo actions (which will also make it easier to refactor later) it's probably best to use a pair of addOrDeleteNoteActions to remove the overwritten note then insert the new one. But wrap them in a macro so they're undone/redone as a single user action.

Ok, that reminds me the comment in master:
` // run twice, delete old note and add new note. this let the undo stack consistent `
even if for some reason this is not happening, and two notes are added

File: src/core/Hydrogen.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (oddtime) [reply src/core/Hydrogen.cpp:2724]: Fantastic!!!

**Comment by oddtime** [comment]
@cme I am currently thinking how to use a macro to bind the `addOrDeleteNoteAction`s, but I don't find a solution.
You meant not to add any new class at all (neither a wrapper class)?

**Comment by cme** [comment]
> @cme I am currently thinking how to use a macro to bind the `addOrDeleteNoteAction`s, but I don't find a solution.
> You meant not to add any new class at all (neither a wrapper class)?

Correct, I mean use `QUndoStack::beginMacro()` before the first and `QUndoStack::endMacro()` after the last of them. That makes `QUndoStack` treat them as a single user operation to undo or redo (and shows up in the history window as a single action)

**Comment by oddtime** [comment]
> Correct, I mean use QUndoStack::beginMacro() before the first and QUndoStack::endMacro() after the last of them. That makes QUndoStack treat them as a single user operation to undo or redo (and shows up in the history window as a single action)

That's wicked! I thought I had to write a macro XD.
Well it seems to work now

File: src/gui/src/HydrogenApp.cpp [hunk 2]
```diff
@@ -636,34 +639,57 @@ void HydrogenApp::onEventQueueTimer()
...
-	while(!pQueue->m_addMidiNoteVector.empty()){
-
-		int rounds = 1;
-		if(pQueue->m_addMidiNoteVector[0].b_noteExist) { // run twice, delete old note and add new note. this let the undo stack consistent 
-			rounds = 2;
+	while( !pQueue->m_addMidiNoteVector.empty() ){
+		Song *pSong = Hydrogen::get_instance()->getSong();
+		Instrument *pSelectedInstrument = pSong->getInstrumentList()->get( pQueue->m_addMidiNoteVector[0].m_row );
+		// find if a (pitch matching) note is already present
+		Note *pOldNote = pSong->getPatternList()->get( Hydrogen::get_instance()->getSelectedPatternNumber() )
```
Comment (theGreatWhiteShark) [inline src/gui/src/HydrogenApp.cpp:646]: Do you use the current pattern over the one stored in `pQueue->m_addMidiNoteVector[0].m_pattern` on purpose?

Comment (theGreatWhiteShark) [reply src/gui/src/HydrogenApp.cpp:646]: Yes

**Comment by theGreatWhiteShark** [comment]
Just tried it. When being in 
1. In a song with two empty patterns selected in bar 1 and 2 respectively
2. in Song mode + loop enabled
3. having one of the patterns selected and shown in the Pattern Editor

Recording notes in the first pattern works fine. But as soon as playback enters the second one a note record results in deleting the corresponding note in the first pattern and adding an additional one in the second. The later will than have multiple equal notes on top of each other.

**Comment by oddtime** [comment]
Ops, I had tried just recording in the first pattern, in pattern mode.
I found other problems in pattern mode as well, recording into the second pattern.
I can't reproduce the same behaviour you describe, because my notes are not deleted in the first pattern (have you built the last commit?) but they are added in the second pattern.

**Comment by oddtime** [comment]
It there are two contemporary patterns in a pattern column (recording in song mode), where should the notes be recorded?
Now they are added to the pattern with biggest index

**Comment by oddtime** [comment]
For me `pQueue->m_addMidiNoteVector[0].m_pattern` as argument of` SE_addOrDeleteNoteAction()`  resolves the problem, because pattern idx in `pOldNote` seems to be always 0:
```
if( pOldNote ) { // note found => remove it
	SE_addOrDeleteNoteAction *action = new SE_addOrDeleteNoteAction( pOldNote->get_position(),
								pOldNote->get_instrument_id(),
								pQueue->m_addMidiNoteVector[0].m_pattern,
								pOldNote->get_length(),
[...]
```

Originally the arguments were all from `pQueue->m_addMidiNoteVector[0]`, is it preferable?

**Comment by theGreatWhiteShark** [comment]
> have you built the last commit?

I experienced this behavior at both 194786f and bdabeaa .

> It there are two contemporary patterns in a pattern column (recording in song mode), where should the notes be recorded?
Now they are added to the pattern with biggest index

@cme updated this in #975

> For me pQueue->m_addMidiNoteVector[0].m_pattern as argument of SE_addOrDeleteNoteAction() resolves the problem, because pattern idx in pOldNote seems to be always 0 (why?):

Hmm. This feels like a bug. I haven't took a close look at this part of the core yet.

**Comment by cme** [comment]
> For me `pQueue->m_addMidiNoteVector[0].m_pattern` as argument of` SE_addOrDeleteNoteAction()` resolves the problem, because pattern idx in `pOldNote` seems to be always 0 (why?):

So... `pattern_idx` seems to only have meaning when notes are associated with a deleted instrument (the deletion of which can be undone). The member doc says:

```
index of the pattern holding this note for undo actions
```

It's a bit tricky since it has the meaning that's obvious from the *name* of the member, it's just that it only applies at in a specific context! :D

Notes don't really know what pattern they belong to, and not by number because that will change if the list gets re-ordered. Since it's only used for undo'ing instrument removal, it would probably be better at some point to remove that field from every Note and use a different structure for the undo operation (maybe, rather than a list of notes, a `multimap< int /*nPattern*/, Note * >`) so the pattern is only stored when it's needed.

In the meantime, using the pattern number from the midi like you have above is the right thing to do :)

(Sorry I didn't notice that when I looked at it before!)

File: src/gui/src/HydrogenApp.cpp [hunk 3]
```diff
@@ -636,34 +639,57 @@ void HydrogenApp::onEventQueueTimer()
...
+																	 pQueue->m_addMidiNoteVector[0].m_column,
+																	 pInstrument,
+																	 pQueue->m_addMidiNoteVector[0].nk_noteKeyVal,
+																	 pQueue->m_addMidiNoteVector[0].no_octaveKeyVal );
+		auto pUndoStack = HydrogenApp::get_instance()->m_pUndoStack;
+		pUndoStack->beginMacro( tr( "Input Midi Note" ) );
+		if( pOldNote ) { // note found => remove it
+			SE_addOrDeleteNoteAction *action = new SE_addOrDeleteNoteAction( pOldNote->get_position(),
+																	 pOldNote->get_instrument_id(),
++																	 pQueue->m_addMidiNoteVector[0].m_pattern,
```
Comment (cme) [inline src/gui/src/HydrogenApp.cpp:657]: Extra '`+`' ;)

**Comment by theGreatWhiteShark** [comment]
> Notes don't really know what pattern they belong to, and not by number because that will change if the list gets re-ordered.

Wicked! This way the name of the member variable is indeed quite misleading.

**Comment by theGreatWhiteShark** [comment]
Not quite sure if this is something meaningful or a false positive but when recording in Input mode set to *Instrument mode* I get a number of errors in the logs.

```
(E) InstrumentList::get idx 23 out of [0;18]
(E) InstrumentList::get idx 19 out of [0;18]
(E) InstrumentList::get idx 23 out of [0;18]
(E) InstrumentList::get idx 21 out of [0;18]
(E) InstrumentList::get idx 19 out of [0;18]
(E) InstrumentList::get idx 21 out of [0;18]
(E) InstrumentList::get idx 19 out of [0;18]
```

Apart from this everything seems to work fine.

**Comment by oddtime** [comment]
seems related to the total number of instruments, but it should appear in _instrument mode_...

**Comment by cme** [comment]
> Not quite sure if this is something meaningful or a false positive but when recording in Input mode set to *Instrument mode* I get a number of errors in the logs.
> 
> ```
> (E) InstrumentList::get idx 23 out of [0;18]
> (E) InstrumentList::get idx 19 out of [0;18]
> (E) InstrumentList::get idx 23 out of [0;18]
> (E) InstrumentList::get idx 21 out of [0;18]
> (E) InstrumentList::get idx 19 out of [0;18]
> (E) InstrumentList::get idx 21 out of [0;18]
> (E) InstrumentList::get idx 19 out of [0;18]
> ```
> 
> Apart from this everything seems to work fine.

Do the notes get inserted at the right instrument & pitch? Does the undo and redo work okay for them?

**Comment by cme** [comment]
> > ```
> > (E) InstrumentList::get idx 23 out of [0;18]
> > (E) InstrumentList::get idx 19 out of [0;18]
> > (E) InstrumentList::get idx 23 out of [0;18]
> > Apart from this everything seems to work fine.
> 
> Do the notes get inserted at the right instrument & pitch? Does the undo and redo work okay for them?

I can reproduce this behaviour on `master` so it's at least unrelated to this PR, and the notes seem to get the right instrument and pitch when inserted.

**Comment by oddtime** [comment]
> Do the notes get inserted at the right instrument & pitch? Does the undo and redo work okay for them?

I confirm it is working. Also, the error messages appears to me if I input notes via the computer keyboard, but not via an external midi keyboard.

**Comment by theGreatWhiteShark** [comment]
Okay. Thanks for looking into it and sorry for the false positive :)

**Review by theGreatWhiteShark — APPROVED** [review]

**Comment by theGreatWhiteShark** [comment]
@oddtime could you merge the latest changes from `master` into this branch so I can merge it?

**Comment by oddtime** [comment]
done

**Comment by theGreatWhiteShark** [comment]
Thanks again! Any preferences which of your PRs I should take a look at next?

**Comment by oddtime** [comment]
I would say #1252 which is super brief,
then #1273 (panL panR deprecation)

**Pull request merged**